### ISSUE TITLE
Session lost upon login with jvmRoute property

### ISSUE THREAD

**Issue by pdudits** [payara/Payara#1447]
# Description #
----------

After user in registered with the session, the session id changes. With jvmRoute property set (to faciliate session stickiness at load balancer), the `Set-Cookie` header is not sent immediately in `Request.addSessionCookie`, rather it is handled by OutputBuffer.addSessionCookieWithJvmRoute.

## Expected Outcome

The first response after login should contain header `Set-Cookie` with newly assigned session id.

## Current Outcome

By the time `OutputBuffer` wants to call `grizzlyResponse.addHeader`, the response is already commited and the call silently fails. User therefore doesn't get any new session cookie, and all subsequent requests are unauthenticated, as the previous session no longer exists.

Observed order of events from adding few breakpoints is following:

1. StandardSession.setId is invoked
2. Header is marked as commited with following stacktrace
  ````
org.glassfish.grizzly.http.HttpHeader.setCommitted(HttpHeader.java:476)
org.glassfish.grizzly.http.HttpCodecFilter.encodeHttpPacket(HttpCodecFilter.java:1570)
org.glassfish.grizzly.http.HttpServerFilter.encodeHttpPacket(HttpServerFilter.java:843)
org.glassfish.grizzly.http.HttpCodecFilter.handleWrite(HttpCodecFilter.java:1449)
org.glassfish.grizzly.filterchain.ExecutorResolver$8.execute(ExecutorResolver.java:111)
org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:890)
org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:858)
org.glassfish.grizzly.http.io.OutputBuffer.flushBuffer(OutputBuffer.java:1059)
org.glassfish.grizzly.http.io.OutputBuffer.flushBinaryBuffers(OutputBuffer.java:1046)
org.glassfish.grizzly.http.io.OutputBuffer.write(OutputBuffer.java:496)
org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:422)
org.apache.catalina.connector.CoyoteWriter.write(CoyoteWriter.java:240)
org.apache.catalina.connector.CoyoteWriter.write(CoyoteWriter.java:249)
com.sun.faces.application.view.WriteBehindStateWriter.write(WriteBehindStateWriter.java:127)
com.sun.faces.renderkit.html_basic.HtmlResponseWriter.endElement(HtmlResponseWriter.java:593)
com.sun.faces.renderkit.html_basic.StylesheetRenderer.endElement(StylesheetRenderer.java:69)
com.sun.faces.renderkit.html_basic.ScriptStyleBaseRenderer.encodeChildren(ScriptStyleBaseRenderer.java:242)
javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:890)
javax.faces.component.UIComponent.encodeAll(UIComponent.java:1856)
com.sun.faces.renderkit.html_basic.HeadRenderer.encodeHeadResources(HeadRenderer.java:113)
com.sun.faces.renderkit.html_basic.HeadRenderer.encodeEnd(HeadRenderer.java:100)
javax.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:920)
javax.faces.component.UIComponent.encodeAll(UIComponent.java:1863)
javax.faces.component.UIComponent.encodeAll(UIComponent.java:1859)
com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:458)
com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:134)
javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:337)
javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:337)
javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:337)
com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:120)
com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:219)
javax.faces.webapp.FacesServlet.service(FacesServlet.java:659)
````
3. OutputBuffer.onCommit 
4. It fails to add the header as the headers are already committed.

## Workaround

Downgrade to grizzly 2.3.25 (the version shipped with 163) helped.

## Steps to reproduce (Only for bug reports) 

1. Start domain
2. Define jvmRoute

     asadmin create-jvm-options --target server-config -DjvmRoute=BOOM

3. Any JSF project with authentication should manifest this behavior.  

### Samples

I'll try to provide a reproducer.

## Environment ##

- **Payara Version**: 4.1.1.171.0.1
- **Edition**: Full
- **JDK Version**: Oracle 8
- **Operating System**: Windows

Another things that may affect the behaviour are:
- JASPIC authentication
- A filter that sets standard security headers

**Comment by lprimak** [comment]
Do you by chance know why it works with 2.3.25?  i.e. what's the key difference?

**Comment by pdudits** [comment]
Not yet, it was just blind attempt at fixing by reverting grizzly modules to previous version.

I'm trying to create a sample project, but it seems that it needs more ingredients than just form auth, jvmRoute and JSF page.

I hopefully get back to it on Monday, and try to look at diffs between grizzly releases.

**Comment by pdudits** [comment]
It appears to be caused by https://github.com/GrizzlyNIO/grizzly-mirror/commit/be8a9c395, as that's where the offending flush was added. (a fix for issue reported by @lprimak, the circle closes ;)). In the meantime the search for reproducing project continues.

**Comment by pdudits** [comment]
I filed the issue for Grizzly: https://java.net/jira/browse/GRIZZLY-1894

**Comment by lprimak** [comment]
Looks like that issue has just been fixed in Grizzly

**Comment by mikecroft** [comment]
Looks like this just missed the 2.3.30 release, so will probably make it in to 173:
https://github.com/javaee/grizzly/commit/35b3977671c154ab81f2039bde25f69e38384f04

**Comment by mikecroft** [comment]
The fix is available in `2.3.31`, released on 2nd June. I created `PAYARA-1731` to upgrade Grizzly to that version for `173`

**Comment by mikecroft** [comment]
fixed by #1655
